hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala

```diff
@@ -0,0 +1,255 @@
+
+  val pe_out_fire = WireInit(0.B)
+  val pe_out_valid = WireInit(0.B)
+  val pe_out_fire_counter = RegInit(0.U(params.sizeConfigWidth.W))
+
+  // reshuffle result data output
+  val reshuffle_output_fire = WireInit(0.B)
+  val write_counter = RegInit(0.U(params.sizeConfigWidth.W))
+
+  // State declaration
+  val sIDLE :: sBUSY :: Nil = Enum(2)
+  val cstate = RegInit(sIDLE)
+  val nstate = WireInit(sIDLE)
+
+  // signals for state transition
+  val config_valid = WireInit(0.B)
+  val computation_finish = WireInit(0.B)
+
+  // Changing states
+  cstate := nstate
+
+  chisel3.dontTouch(cstate)
+  switch(cstate) {
+    is(sIDLE) {
+      when(config_valid) {
+        nstate := sBUSY
+      }.otherwise {
+        nstate := sIDLE
+      }
+
+    }
+    is(sBUSY) {
+      when(computation_finish) {
+        nstate := sIDLE
+      }.otherwise {
+        nstate := sBUSY
+      }
+    }
+  }
+
+  io.busy_o := cstate === sBUSY
+
+  val performance_counter = RegInit(0.U(32.W))
+  when(cstate === sBUSY) {
+    performance_counter := performance_counter + 1.U
+  }.elsewhen(config_valid) {
+    performance_counter := 0.U
+  }
+  io.performance_counter := performance_counter
+
+  config_valid := io.ctrl.fire
+
+  // when config valid, store the configuration for later computation
+  when(config_valid) {
+    ctrl_csr.opcode_i := io.ctrl.bits(0)(1, 0)
+    ctrl_csr.reduceLen_i := io.ctrl.bits(0)(6, 2)
+    ctrl_csr.tloop2Len_i := io.ctrl.bits(0)(31, 7)
+  }
+
+  // when opcode is 2, do MAXPool
+  val doMAXPool = ctrl_csr.opcode_i === 2.U && cstate === sBUSY
+
+  // -------------------------------------------------------------------
+  // pool simd logic
+  // -------------------------------------------------------------------
+
+  // give each MAXPoolPE right control signal and data
+  // collect the result of each MAXPoolPE
+  when(doMAXPool) {
+    for (i <- 0 until params.poolLaneLen) {
+      // init signal for the first MAXPoolPE
+      lane(i).io.ctrl.init_i := pe_input_fire_counter === 0.U
+
+      lane(i).io.data.input_i.bits := io.data.input_i
+        .bits(
+          (i + 1) * params.inputWidth - 1,
+          (i) * params.inputWidth
+        )
+        .asSInt
+      lane(
+        i
+      ).io.data.input_i.valid := io.data.input_i.valid && io.data.input_i.ready
+
+      result(i) := lane(i).io.data.output_o.bits
+      lane(i).io.data.output_o.ready := Mux(
+        io.data.output_o.valid,
+        io.data.output_o.ready,
+        1.B
+      )
+    }
+  }.otherwise {
+    lane.map(_.io.ctrl.init_i := false.B)
+    lane.map(_.io.data.input_i.bits := 0.S)
+    lane.map(_.io.data.input_i.valid := false.B)
+    lane.map(_.io.data.output_o.ready := false.B)
+    result.map(_ := 0.S)
+  }
+
+  // for generating the init signal for the first MAXPoolPE
+  pe_input_fire := lane.map(_.io.data.input_i.fire).reduce(_ && _)
+  when(
+    pe_input_fire && pe_input_fire_counter =/= (ctrl_csr.reduceLen_i - 1.U) && cstate === sBUSY
```

Comment:
Is there some kind of POSIX ordering here? Or is Chisel smart enough to know which to evaluate first? Like would it be better if:

`pe_input_fire && (pe_input_fire_counter =/= (ctrl_csr.reduceLen_i - 1.U)) && (cstate === sBUSY)`

Reply:
It is